src/TON/Address.cpp

```diff
@@ -0,0 +1,210 @@
+    }
+    return isValid;
+}
+
+bool Address::parseAddress(const std::string& addressStr_in, Address& addr_inout) {
+    // try several formats, start with the common one, stop if one matches
+    bool isValidUser = parseUserAddress(addressStr_in, addr_inout);
+    if (isValidUser) {
+        return true;
+    }
+    bool isValidRaw = parseRawAddress(addressStr_in, addr_inout);
+    return isValidRaw;
+}
+
+bool Address::parseRawAddress(const std::string& addressStr_in, Address& addr_inout) {
+    // split by colon ':'
+    auto colidx = addressStr_in.find(':');
+
+    if (colidx == std::string::npos) {
+        // no colon, invalid
+        return false;
+    }
+    if (colidx < 1 || colidx >= addressStr_in.length() - 1) {
+        // colon in wrong position
+        return false;
+    }
+    std::string workchainStr = addressStr_in.substr(0, colidx);
+    std::string addressStr = addressStr_in.substr(colidx + 1, addressStr_in.length() - colidx - 1);
+
+    WorkchainId_t workchainId;
+    try {
+        workchainId = std::stoi(workchainStr);
+    } catch (const std::exception& e) {
+        // workchain ID is invalid (not a decimal number)
+        return false;
+    }
+    if (!Workchain::isValid(workchainId)) {
+        // invalid workchain ID
+        return false;
+    }
+    addr_inout.workchainId = workchainId;
+
+    if (addressStr.length() != AddressLength * 2) {
+        // wrong length of address part
+        return false;
+    }
+
+    addr_inout.addrBytes = parse_hex(addressStr);
+
+    addr_inout.isBounceable = true;
+    addr_inout.isTestOnly = false;
+
+    return true;
+}
+
+bool Address::parseUserAddress(const std::string& addressStr_in, Address& addr_inout) {
+    Data bytes;
+    try {
+        bytes = Base64::decodeBase64Url(addressStr_in);
+    } catch (const std::exception& ex) {
+        return false;
+    }
+
+    // check length
+    if (bytes.size() != 36) {
+        return false;
+    }
+    assert(bytes.size() >= 36);
+
+    addr_inout.isBounceable = true;
+    addr_inout.isTestOnly = false;
+    byte tagByte = bytes[0];
+    if (tagByte >= 0x80) {
+        // test-only
+        tagByte -= 0x80;
+        addr_inout.isTestOnly = true;
+    }
+    if (tagByte >= 0x40) {
+        // not bounceable
+        tagByte -= 0x40;
+        addr_inout.isBounceable = false;
+    }
+    if (tagByte != 0x11) {
+        // invalid tag
+        return false;
+    }
+
+    byte chainId = bytes[1];
+    switch (chainId) {
+    case 0x00:
+        addr_inout.workchainId = Workchain::BasicChainId;
+        break;
+    case 0xff:
+        addr_inout.workchainId = Workchain::MasterChainId;
+        break;
+    default:
+        return false; // invalid chain
+    }
+
+    // 32 bytes address
+    addr_inout.addrBytes = Data(AddressLength);
+    std::copy(bytes.begin() + 2, bytes.begin() + 2 + AddressLength, addr_inout.addrBytes.begin());
+
+    // check CRC
+    uint16_t crcGiven = static_cast<uint16_t>(bytes[34] << 8) + bytes[35];
```

Comment:
we can extract a computeChecksum method

Reply:
One place, one line usage, I leave here.